# meth-pops
Python tool for visualizing methylation status as lollipop diagram.

**Note: It only supports .sam/.bam file generated by running [Bismark](https://github.com/FelixKrueger/Bismark).**

## Examples

**Lollipop diagram of CpG sites.**

![lollipop](./img/lollipop.png)

**Commandline visualization of  methylation status of CpGs within specified genomic region.** 

![commandline](./img/commandline.gif)

**Pretty-print.**

![pretty_print](./img/pretty_print.gif)

## Usage

Clone or download ZIP of this repository, and run the **methpops** package with python.

```shell
$PY methpops --help
```

When you type the command above, you'll see some help messages as below. (Note that `$PY` should be replaced with appropriate path to the python executable.)

```shell
usage: methpops [-h] -i INPUT [-o OUTPUT] [-d TEMPDIR] [-t THREADS] [-c]
                [--pretty] [-l] [-r REGION]

optional arguments:
  -h, --help            show this help message and exit
  -i INPUT, --input INPUT
                        Input file.
  -o OUTPUT, --output OUTPUT
                        Output lollipop plot.
  -d TEMPDIR, --tempdir TEMPDIR
                        Temporary directory in which the .sorted.bam file and
                        bam index file will be saved.
  -t THREADS, --threads THREADS
                        Number of threads to be used.
  -c, --commandline     Show methylation status at commandline.
  --pretty              Show methylation status at commandline in prettier
                        way.
  -l, --lollipop        Save lollipop plot. Default: True
  -r REGION, --region REGION
                        Genomic region to visualize.
```

This might help you run the program.